app/uk/gov/hmrc/eusubsidycompliancefrontend/controllers/SubsidyController.scala

```diff
@@ -58,66 +64,52 @@ class SubsidyController @Inject()(
-      case (None, _, Some(undertaking)) => // initialise the empty Journey model
-        Ok(reportPaymentPage(None, undertaking))
-      case _ => handleMissingSessionData("Subsidy journey")
+
+    val result = for {
+      _           <- store.get[SubsidyJourney].toContext.orElse(store.put(SubsidyJourney()).toContext)
+      undertaking <- store.get[Undertaking].toContext
+      reference   <- undertaking.reference.toContext
+    } yield (reference, undertaking)
+
+    result.foldF(handleMissingSessionData("Subsidy journey")) {
+      case (reference, undertaking) =>
+        retrieveSubsidiesOrNone(reference).map { subsidies =>
+          Ok(reportPaymentPage(
+            subsidies,
+            undertaking,
+            timeProvider.today.toEarliestTaxYearStart,
+            timeProvider.today.toTaxYearEnd.minusYears(1),
+            timeProvider.today.toTaxYearStart,
+          ))
+        }
     }
   }
 
+  private def retrieveSubsidiesOrNone(r: UndertakingRef)(implicit hc: HeaderCarrier) =
+    escService
+      .retrieveSubsidy(SubsidyRetrieve(r, None))
+      .map(Option(_))
+      .fallbackTo(Option.empty.toFuture)
+
   def postReportPayment: Action[AnyContent] = escAuthentication.async { implicit request =>
     implicit val eori: EORI = request.eoriNumber
     reportPaymentForm.bindFromRequest().fold(
-      _ => throw new IllegalStateException("value hard-coded, form hacking?"),
-      form => {
-        for {
-          journey <- store.update[SubsidyJourney]({
-            _.map { y =>
-              y.copy(reportPayment = y.reportPayment.copy(value = Some(form.value.toBoolean)))
-            }
-          })
-          redirect <- getJourneyNext(journey)
-        } yield redirect
-      }
+      _ => throw new IllegalStateException("report payment form submission failed"),
+      (form: FormValues) => for {
+        journey <- store.update[SubsidyJourney](updateReportPayment(form))
+        redirect <- getJourneyNext(journey)
+      } yield redirect
     )
   }
 
   def getClaimAmount: Action[AnyContent] = escAuthentication.async { implicit request =>
     implicit val eori: EORI = request.eoriNumber
-    //TODO add 'getPrevious to all'
+    // TODO - add 'getPrevious to all'
     store.get[SubsidyJourney].flatMap {
-      case Some(journey) =>
-        journeyTraverseService.getPrevious[SubsidyJourney].flatMap { previous =>
-          journey
-            .claimAmount
-            .value
-            .fold(
-              Future.successful(
-                Ok(addClaimAmountPage(claimAmountForm, previous))
-              )
-            ) { x =>
-              Future.successful(
-                Ok(addClaimAmountPage(claimAmountForm.fill(x), previous))
-              )
-            }
-
-        }
+      case Some(journey) => journeyTraverseService.getPrevious[SubsidyJourney].flatMap { p =>
+        val form = journey.claimAmount.value.fold(claimAmountForm)(claimAmountForm.fill)
+        Ok(addClaimAmountPage(form, p)).toFuture
```

Comment:
minor: If change flatmap on line 109 to map , we won't need toFuture here

Reply:
Done